apis/k8ssandra/v1alpha1/k8ssandracluster_webhook.go

```diff
@@ -192,6 +174,71 @@ func (r *K8ssandraCluster) ValidateUpdate(old runtime.Object) (admission.Warning
 	return nil, nil
 }
 
+func validateUpdateNumTokens(
+	oldCassandra *CassandraClusterTemplate,
+	newCassandra *CassandraClusterTemplate,
+) error {
+	oldNumTokensPerDc, err := numTokensPerDc(oldCassandra)
+	if err != nil {
+		return err
+	}
+	newNumTokensPerDc, err := numTokensPerDc(newCassandra)
+	if err != nil {
+		return err
+	}
+
+	for dcName, newNumTokens := range newNumTokensPerDc {
+		oldNumTokens, oldExists := oldNumTokensPerDc[dcName]
+		if oldExists && oldNumTokens != newNumTokens {
+			return ErrNumTokens
+		}
+	}
+	return nil
+}
+
+func numTokensPerDc(cassandra *CassandraClusterTemplate) (map[string]interface{}, error) {
```

Comment:
question: Why interface{} and not map[string]float64. Or actually just map[string]uint / int, since we don't use floats for anything in reality (it's a YAML/JSON problem) and we should throw an error if we can't map to uint value.

Reply:
Because it's good enough for the task at hand, which is only about comparing values.

We could indeed check that the values are integers, but then why stop at `num_tokens`, and not have the webhook validate every other config value? It's not the approach we have taken so far, so I assume we want to keep things as light as possible here and do more advanced checks later (in this case `computeDcNumTokens()`).

As to float64 for the defaults, it's because that's what unmarshaling into `Unstructured` produces for any numeric value:
```go
	u := new(Unstructured)
	u.UnmarshalJSON([]byte(`{"num_token": 16}`))
	numToken, _ := u.Get("num_token")
	i16 := 16
	f16 := float64(i16)
	fmt.Println("numToken==i16", numToken == i16) // false
	fmt.Println("numToken==f16", numToken == f16) // true
```

Reply:
> As to float64 for the defaults, it's because that's what unmarshaling into Unstructured produces for any numeric value

It is, because JSON & YAML do not accept any other numeric value. Problem itself is more interesting if you compare:

"num_tokens": "16" and "num_tokens": 16. Both are the same value, but they could be unmarshalled differently. We do not force users to use : 16 or : "16" and most users wouldn't even able to tell what the difference is. From your example, we end up with a difference however:

```go
func TestUnmarshalNumbers(t *testing.T) {
	dest := make(map[string]interface{}, 1)
	assert.NoError(t, json.Unmarshal([]byte(`{"num_token": "16"}`), &dest))
	numToken := dest["num_token"]
	i16 := 16
	f16 := float64(i16)

	assert.NotEqual(t, numToken, i16)
	assert.NotEqual(t, numToken, f16)
}
```

Neither compare will work. One needs to parse it first as number (using strconv) and only then compare the resulting int.

Reply:
True, this PR assumes the user will provide a YAML number (as does the current implementation on the main branch btw).
If supporting YAML strings is also a requirement, then it would make sense to parse.